syne_tune/optimizer/schedulers/transfer_learning/bounding_box.py

```diff
@@ -0,0 +1,118 @@
+from syne_tune.search_space import Categorical
+
+
+class BoundingBox(TransferLearningScheduler):
+    def __init__(
+            self,
+            scheduler_fun: Callable[[Dict, str, str], TrialScheduler],
+            config_space: Dict,
+            metric: str,
+            transfer_learning_evaluations: Dict[str, TransferLearningTaskEvaluations],
+            mode: Optional[str] = None,
+            num_hyperparameters_per_task: int = 1,
+    ):
+        """
+        Simple baseline that computes a bounding-box of the best candidate found in previous tasks to restrict the
+         search space to only good candidates. The bounding-box is obtained by restricting to the min-max of best
+         numerical hyperparameters and restricting to the set of best candidates on categorical parameters.
+
+        Reference: Learning search spaces for Bayesian optimization: Another view of hyperparameter transfer learning.
+        Valerio Perrone, Huibin Shen, Matthias Seeger, Cédric Archambeau, Rodolphe Jenatton. Neurips 2019.
+
+        :param scheduler_fun: function that takes a configuration space (Dict), a mode (str) and a metric (str)
+        and returns a scheduler. This is required since the final configuration space is known only after computing
+        a bounding-box. For instance,
+        `scheduler_fun=lambda new_config_space, mode, metric: RandomSearch(new_config_space, metric, mode)`
+        will consider a random-search on the config-space is restricted to the bounding of best evaluations of previous
+        tasks.
+        :param config_space: initial search-space to consider, will be updated to the bounding of best evaluations of
+        previous tasks
+        :param metric: objective name to optimize, must be present in transfer learning evaluations.
+        :param transfer_learning_evaluations: dictionary from task name to offline evaluations.
+        :param mode: mode to be considered, default to min.
+        :param num_hyperparameters_per_task: number of best hyperparameter to take per task when computing the bounding
+        box, default to 1.
+        """
+        self.mode = mode if mode is not None else "min"
+
+        new_config_space = self.compute_box(
+            config_space=config_space,
+            transfer_learning_evaluations=transfer_learning_evaluations,
+            mode=self.mode,
+            num_hyperparameters_per_task=num_hyperparameters_per_task,
+            metric=metric
+        )
+        self.config_space = new_config_space
+        print(f"hyperparameter ranges of best previous configurations {new_config_space}")
+        print(f"({sp.search_space_size(new_config_space)} options)")
+        self.scheduler = scheduler_fun(new_config_space, mode, metric)
+        super(BoundingBox, self).__init__(
+            config_space=new_config_space,
+            transfer_learning_evaluations=transfer_learning_evaluations,
+            metric_names=self.scheduler.metric_names(),
+        )
```

Comment:
The config space is not known before computing the bounding-box so I am not sure how init can be called first (unless we update the field `config_space`).

Reply:
good point. but atm the transfer learning class only provides this if the search space is well-defined. as far as I understand this correctly, it will fail in this case before ever reaching this point. no strong preference in this case, but I'd probably overwrite config_space later to make use of the test logic.

Reply:
I am not sure how to overwrite the config-space safely so I propose to let it this way if this works for you.

Reply:
We cannot overwrite or change config_space after calling the superclass init. That one will initialize other members based on configspace, it will create a searcher, HyperparameterRanges.

The pattern of just changing members of a class is not good. Such members should either be read-only, or only modified by methods of the class.

Same with Domain class in search_ranges.py, should not just change upper, lower. That one is still managable maybe, but with a general scheduler, you have just no idea what is done with config_space at construction.

I am still most in favor of a factory which then just creates the modified scheduler, but let you decide. But please do not do things like overwriting members of classes, this is a pattern that leads to big problems.

I am in general trying to use _ for such members, but it can make sense to make them visible read-only. To be safe, we could also use _ and then a property, but we often don't.

Thanks.

Reply:
David could make the test in the superclass a method and then call it before creating new_config_space. That would make everybody happy.

Reply:
> But please do not do things like overwriting members of classes, this is a pattern that leads to big problems.

To be clear currently there is no member overwriting in the PR and I am also saying that we cannot modify the class to override the config-space after creating the scheduler.

> David could make the test in the superclass a method and then call it before creating new_config_space. That would make everybody happy.

Are you suggesting this?
```python
class BoundingBox(TransferLearningScheduler):
    def __init__(..., transfer_learning_evaluations)
       self._check_consistency(transfer_learning_evaluations) # <- check before init
       new_config_space = self.compute_box(...)
       ...
       super(BoundingBox, self).__init__(...)

Reply:
Yes. You can also call this internal method in the superclass constructor, to be safe. It would then be called 2x.

Reply:
+1 to property and "_". It wasn't clear to me that this would cause problems

I'm fine this solution of calling the check.

Reply:
Ok, I sent a revision that performs the check before computing the BB.